File: packages/python/trikhub/langchain/schema_converter.py
"""
JSON Schema → Pydantic model converter.

Converts JSON Schema objects (from manifest inputSchema/outputSchema)
into Pydantic models for use with LangChain StructuredTool.

Mirrors packages/js/gateway/src/langchain/schema-converter.ts
"""

from __future__ import annotations

from typing import Any

from pydantic import Field, create_model


def json_schema_to_pydantic(
    schema: dict[str, Any], model_name: str = "DynamicModel"
) -> type:
    """
    Convert a JSON Schema definition to a Pydantic model class.

    Supports: string (maxLength, enum), number, integer, boolean,
    object with properties + required, and description passthrough.

    Raises ValueError on unsupported constructs ($ref, oneOf, anyOf, allOf, arrays).
    """
    # Enum at top level
    if "enum" in schema:
        values = schema["enum"]
        if not values:
            raise ValueError("json_schema_to_pydantic: empty enum is not supported")
        # Return a model with a single 'value' field constrained to enum values
        # For top-level enums, we create a Literal type
        from typing import Literal

        literal_type = Literal[tuple(str(v) for v in values)]  # type: ignore[valid-type]
        field_kwargs: dict[str, Any] = {}
        if schema.get("description"):
            field_kwargs["description"] = schema["description"]
        return create_model(
            model_name,
            value=(literal_type, Field(**field_kwargs)),
        )

    schema_type = schema.get("type")

    if schema_type == "object":
        return _build_object_model(schema, model_name)

    # For non-object types, wrap in a single-field model
    field_type, field_info = _build_field(schema)
    return create_model(model_name, value=(field_type, field_info))


def json_schema_to_field(schema: dict[str, Any]) -> tuple[type, Any]:
    """
    Convert a JSON Schema property into a (type, Field) tuple
    suitable for use in Pydantic create_model().
    """
    return _build_field(schema)


def _build_object_model(
    schema: dict[str, Any], model_name: str = "DynamicModel"
) -> type:
    """Build a Pydantic model from an object-type JSON Schema."""
    properties = schema.get("properties")
    if not properties:
        # Object with no properties — use a plain dict
        field_kwargs: dict[str, Any] = {}
        if schema.get("description"):
            field_kwargs["description"] = schema["description"]
        return create_model(
            model_name,
            __root__=(dict[str, Any], Field(default_factory=dict, **field_kwargs)),
        )

    required_set = set(schema.get("required", []))
    fields: dict[str, Any] = {}

    for key, prop_schema in properties.items():
        field_type, field_info = _build_field(prop_schema)
        if key not in required_set:
            # Optional field with None default
            field_type = field_type | None  # type: ignore[assignment]
            field_info = Field(default=None, description=field_info.description)
        fields[key] = (field_type, field_info)

    model = create_model(model_name, **fields)
    if schema.get("description"):
        model.__doc__ = schema["description"]
    return model


def _build_field(schema: dict[str, Any]) -> tuple[type, Any]:
    """Convert a single JSON Schema property to a (type, Field) tuple."""
    # Enum
    if "enum" in schema:
        values = schema["enum"]
        if not values:
            raise ValueError("json_schema_to_pydantic: empty enum is not supported")
        from typing import Literal

        literal_type = Literal[tuple(str(v) for v in values)]  # type: ignore[valid-type]
        kwargs: dict[str, Any] = {}
        if schema.get("description"):
            kwargs["description"] = schema["description"]
        return literal_type, Field(**kwargs)  # type: ignore[return-value]

    schema_type = schema.get("type")

    if schema_type == "string":
        return _build_string_field(schema)
    if schema_type in ("number", "integer"):
        return _build_number_field(schema)
    if schema_type == "boolean":
        kwargs = {}
        if schema.get("description"):
            kwargs["description"] = schema["description"]
        return bool, Field(**kwargs)
    if schema_type == "object":
        # Nested object — build a sub-model
        sub_model = _build_object_model(schema, "NestedModel")
        kwargs = {}
        if schema.get("description"):
            kwargs["description"] = schema["description"]
        return sub_model, Field(**kwargs)

    # Unsupported constructs
    if "$ref" in schema:
        raise ValueError("json_schema_to_pydantic: $ref is not supported")
    if schema_type == "array":
        raise ValueError("json_schema_to_pydantic: array type is not supported")
    if isinstance(schema_type, list):
        raise ValueError("json_schema_to_pydantic: union types are not supported")

    raise ValueError(
        f'json_schema_to_pydantic: unsupported schema type "{schema_type}"'
    )


def _build_string_field(schema: dict[str, Any]) -> tuple[type, Any]:
    """Build a string field from JSON Schema."""
    kwargs: dict[str, Any] = {}

    if schema.get("maxLength") is not None:
        kwargs["max_length"] = schema["maxLength"]
    if schema.get("minLength") is not None:
        kwargs["min_length"] = schema["minLength"]
    if schema.get("pattern"):
        kwargs["pattern"] = schema["pattern"]

    desc = schema.get("description", "")
    if schema.get("format"):
        format_hint = f"(format: {schema['format']})"
        desc = f"{desc} {format_hint}" if desc else format_hint
    if desc:
        kwargs["description"] = desc

    return str, Field(**kwargs)


def _build_number_field(schema: dict[str, Any]) -> tuple[type, Any]:
    """Build a number/integer field from JSON Schema."""
    is_integer = schema.get("type") == "integer"
    field_type: type = int if is_integer else float
    kwargs: dict[str, Any] = {}

    if schema.get("minimum") is not None:
        kwargs["ge"] = schema["minimum"]
    if schema.get("maximum") is not None:
        kwargs["le"] = schema["maximum"]
    if schema.get("description"):
        kwargs["description"] = schema["description"]

    return field_type, Field(**kwargs)

File: packages/js/gateway/src/langchain/schema-converter.ts
/**
 * JSON Schema → Zod schema converter.
 *
 * Converts JSON Schema objects (from manifest inputSchema/outputSchema)
 * into Zod schemas for use with LangChain DynamicStructuredTool.
 */

import { z } from 'zod';
import type { JSONSchema } from '@trikhub/manifest';

/**
 * Convert a JSON Schema definition to a Zod schema.
 *
 * Supports: string (maxLength, enum, pattern, format), number, integer,
 * boolean, object with properties + required, and description passthrough.
 *
 * Throws on unsupported constructs ($ref, oneOf, anyOf, allOf, arrays, etc.).
 */
export function jsonSchemaToZod(schema: JSONSchema): z.ZodTypeAny {
  // Enum at top level (applies to any type)
  if (schema.enum) {
    if (schema.enum.length === 0) {
      throw new Error('jsonSchemaToZod: empty enum is not supported');
    }
    const values = schema.enum.map((v) => String(v));
    let zodEnum = z.enum(values as [string, ...string[]]);
    if (schema.description) {
      zodEnum = zodEnum.describe(schema.description);
    }
    return zodEnum;
  }

  const type = schema.type;

  if (type === 'string') {
    return buildStringSchema(schema);
  }

  if (type === 'number' || type === 'integer') {
    return buildNumberSchema(schema);
  }

  if (type === 'boolean') {
    let zodBool: z.ZodTypeAny = z.boolean();
    if (schema.description) {
      zodBool = zodBool.describe(schema.description);
    }
    return zodBool;
  }

  if (type === 'object') {
    return buildObjectSchema(schema);
  }

  // Unsupported constructs
  if (schema.$ref) {
    throw new Error('jsonSchemaToZod: $ref is not supported');
  }
  if (type === 'array') {
    throw new Error('jsonSchemaToZod: array type is not supported');
  }
  if (Array.isArray(type)) {
    throw new Error('jsonSchemaToZod: union types are not supported');
  }

  throw new Error(`jsonSchemaToZod: unsupported schema type "${type ?? 'undefined'}"`);
}

function buildStringSchema(schema: JSONSchema): z.ZodTypeAny {
  let zodStr = z.string();

  if (schema.maxLength !== undefined) {
    zodStr = zodStr.max(schema.maxLength);
  }

  if (schema.minLength !== undefined) {
    zodStr = zodStr.min(schema.minLength);
  }

  if (schema.pattern) {
    zodStr = zodStr.regex(new RegExp(schema.pattern));
  }

  // Format → basic string with description hint (Zod doesn't have native format)
  // We add the format info to the description for LLM guidance
  let desc = schema.description ?? '';
  if (schema.format) {
    const formatHint = `(format: ${schema.format})`;
    desc = desc ? `${desc} ${formatHint}` : formatHint;
  }

  let result: z.ZodTypeAny = zodStr;
  if (desc) {
    result = result.describe(desc);
  }
  return result;
}

function buildNumberSchema(schema: JSONSchema): z.ZodTypeAny {
  const isInteger = schema.type === 'integer';
  let zodNum = isInteger ? z.number().int() : z.number();

  if (schema.minimum !== undefined) {
    zodNum = zodNum.min(schema.minimum);
  }

  if (schema.maximum !== undefined) {
    zodNum = zodNum.max(schema.maximum);
  }

  let result: z.ZodTypeAny = zodNum;
  if (schema.description) {
    result = result.describe(schema.description);
  }
  return result;
}

function buildObjectSchema(schema: JSONSchema): z.ZodTypeAny {
  const properties = schema.properties;
  if (!properties) {
    // Object with no properties — use record
    let zodRecord: z.ZodTypeAny = z.record(z.unknown());
    if (schema.description) {
      zodRecord = zodRecord.describe(schema.description);
    }
    return zodRecord;
  }

  const requiredSet = new Set(schema.required ?? []);
  const shape: Record<string, z.ZodTypeAny> = {};

  for (const [key, propSchema] of Object.entries(properties)) {
    let field = jsonSchemaToZod(propSchema);
    if (!requiredSet.has(key)) {
      field = field.optional();
    }
    shape[key] = field;
  }

  let zodObj: z.ZodTypeAny = z.object(shape);
  if (schema.description) {
    zodObj = zodObj.describe(schema.description);
  }
  return zodObj;
}
